core/src/sumeragi/message.rs

```diff
@@ -143,15 +139,49 @@ impl From<CommittedBlock> for BlockCommitted {
 #[non_exhaustive]
 pub struct BlockSyncUpdate {
     /// The corresponding block.
-    pub block: VersionedCandidateCommittedBlock,
+    block: VersionedCommittedBlock,
 }
 
 impl From<VersionedCommittedBlock> for BlockSyncUpdate {
     fn from(block: VersionedCommittedBlock) -> Self {
-        Self {
-            block: block.into(),
+        Self { block }
+    }
+}
+
+impl BlockSyncUpdate {
+    /// Extract block from block sync update message.
+    ///
+    /// # Errors
+    /// - When the block is invalid.
+    pub fn validate_and_extract_block(
+        self,
+        transaction_validator: &TransactionValidator,
+        wsv: &WorldStateView,
+        latest_block: Option<HashOf<VersionedCommittedBlock>>,
+        block_height: u64,
+    ) -> Result<VersionedCommittedBlock, eyre::Report> {
+        if block_height == 0 {
```

Comment:
Initial idea behind `const generic` was that we know exact places in code where we expect block with certain height (0 or more than zero) and know we fall back to runtime check.

Reply:
I have created this issue, https://github.com/hyperledger/iroha/issues/3372.

Reply:
I mean that we still can use `const generics`, not that we should remove them. Because we only expect genesis block inside wait for genesis loop, no?
I don't think that soft fork on genesis block is possible.

Reply:
The genesis wait loop calls this function because it recieves block sync messages.

Reply:
Should I create a seperate version of the extract function for use by the genesis code only?